syntax = "proto3";

package common;

import "google/protobuf/timestamp.proto";

// Status codes following gRPC conventions
enum StatusCode {
    OK = 0;
    CANCELLED = 1;
    UNKNOWN = 2;
    INVALID_ARGUMENT = 3;
    DEADLINE_EXCEEDED = 4;
    NOT_FOUND = 5;
    ALREADY_EXISTS = 6;
    PERMISSION_DENIED = 7;
    RESOURCE_EXHAUSTED = 8;
    FAILED_PRECONDITION = 9;
    ABORTED = 10;
    OUT_OF_RANGE = 11;
    UNIMPLEMENTED = 12;
    INTERNAL = 13;
    UNAVAILABLE = 14;
    DATA_LOSS = 15;
    UNAUTHENTICATED = 16;
    PARTIAL_SUCCESS = 17;
}

// Status message for responses
message Status {
    StatusCode code = 1;
    string message = 2;
    repeated StatusDetail details = 3;
    google.protobuf.Timestamp timestamp = 4;
}

// Detailed status information
message StatusDetail {
    string type = 1;
    string field = 2;
    string description = 3;
    map<string, string> metadata = 4;
}

// Error details
message ErrorInfo {
    string reason = 1;
    string domain = 2;
    map<string, string> metadata = 3;
}

// Retry information
message RetryInfo {
    google.protobuf.Timestamp retry_after = 1;
    int32 retry_count = 2;
    int32 max_retries = 3;
}

// Debug information
message DebugInfo {
    repeated string stack_entries = 1;
    string detail = 2;
}
